### PR TITLE
Add opt-in Apple Core AI (.aimodel) backend for iOS 27+

### PR DESCRIPTION
**Pull request by glenn-jocher**
Adds an opt-in Core AI backend to the `UltralyticsYOLO` SDK and app. Core ML remains the default: official model IDs, bare bundle names, the bundled models and the remote model list behave exactly as on `main`. SDK users opt in by passing an `.aimodel` path or an `.aimodel.zip` URL; app users opt in with a Settings toggle that is off by default. Deployment targets, Swift tools 5.10 and the existing public API are unchanged; `CoreAI.framework` is weak-linked.

```swift
let model = YOLO(".../yolo26n.aimodel", task: .detect) { result in /* ... */ }  // iOS 27+ device
```

## What changed

- **Backend** (`CoreAIRequest.swift`): `.aimodel` loads through Core AI behind `#if canImport(CoreAI)` and `#available(iOS 27.0, *)`. `useGpu` keeps its meaning of hardware acceleration: when true Core AI places the model across the Neural Engine, GPU and CPU, when false it is pinned to CPU. An `.aimodel` on an OS or SDK without Core AI fails with `PredictorError.coreAIUnavailable`. New public `BasePredictor.isCoreAIAvailable` lets an opt-in app decide which file to ship.
- **Seam**: `visionRequest` is a `VNRequest`; a `CoreAIRequest` subclass carries Float32 `MLMultiArray` copies of the outputs, and one `featureArrays(_:)` accessor replaces the duplicated Vision casts and `firstFeatureArray`. The 7 task decoders are otherwise unchanged and pick the head by output shape.
- **Preprocessing**: Swift letterbox (center-crop for classify) with the same geometry as the existing un-letterbox math, 114 gray padding, vImage BGRA to RGB CHW, real `preMs` for `predictOnImage`. The input tensor is allocated once and filled in bulk (`NDArray(scalars:)` cost about 55 ms per 640x640 frame on the device).
- **Load recovery**: on iOS 27.0 a stale specialization cache entry (asset replaced under the same path, or app reinstalled) fails the load with `_GenericObjCError.nilError` instead of being rebuilt; the loader evicts the asset's cache entries and retries once. Six of the seven nano assets hit this on the test device and recovered.
- **Metadata**: labels come from the asset's `metadata.json`, read as plain JSON and parsed by the existing `parseLabels`.
- **Explicit `.aimodel` routes**: the path resolver accepts an `.aimodel` path or a bundle name carrying the `.aimodel` extension (bare names resolve Core ML exactly as before); the downloader and cache accept `.aimodel.zip` URLs (`metadata.json` validity marker, no compile step). 
- **App opt-in**: Settings → Ultralytics YOLO → "Core AI Models (iOS 27+)" (`use_core_ai`, default off), next to the existing external-display toggle. When it is on and Core AI is available, the model list uses the official `.aimodel.zip` assets and any bundled `.aimodel`; when it is off, or on earlier iOS, the app behaves exactly as on `main`. The two formats cache side by side and neither ever replaces the other. When the setting changed, returning to the foreground relists the models in the selected format and loads the first model of the current task; when it did not change, foregrounding does nothing, as on `main`. A flip during a download cannot misplace the model being installed.
- **Classify**: Core ML raw-tensor results keep `main`'s softmax; Core AI outputs are already probabilities and skip it.
- **Deleted**: the app's in-memory `MLModel` cache, which was never consumed and whose `MLModel?` completions cannot represent a Core AI install.
- **Scripts**: `export-models.py` exports Core ML by default; `--formats coreai` exports and verifies the official opt-in assets (raw head for detect, segment, pose and OBB).
- Docs in both languages, `docs/performance.md` with the apples-to-apples on-device tables, version 8.9.15.

Size: code-only +539/−188. The Core AI backend file is the bulk; the predictor seam is net-negative, and the benchmark harness used for the results was removed after the runs (source in ultralytics/yolo-ios-app#320).

Verified: simulator `clean build test` 110 tests, 0 failures (includes a real preprocessing test for RGB order, row order and pad geometry); SDK scheme and app build for `-sdk iphoneos`; strict Periphery clean. On an iPhone 17 Pro (iOS 27.0): all 7 tasks correct against Core ML, live camera checked for all tasks, detect s/m/l/x and the x model of every other task load and run, a 22,000-inference soak did not reproduce apple/coreai-torch#75. Core AI cannot run on macOS 26 or in the iOS Simulator, so CI exercises the Core ML path and the preprocessing only.

## Coordinated PRs

Core ML (`.mlpackage`) remains the default everywhere. Core AI (`.aimodel`) becomes an opt-in for iOS 27 and later devices: an app passes an `.aimodel` path or an `.aimodel.zip` URL. There is no new flag or argument; the file extension selects the backend. Core AI is not available on earlier iOS versions or in the iOS Simulator. Tracking issue for a future default flip: ultralytics/yolo-ios-app#320.

| PR | Scope | Merge order |
| --- | --- | --- |
| ultralytics/ultralytics#26230 | Core AI asset description, docs for the SDK opt-in and the raw-head recipe | 3 |
| **this PR** | Opt-in Core AI backend in the `UltralyticsYOLO` SDK and app, export scripts, on-device benchmarks | 1 |
| ultralytics/yolo-flutter-app#596 | Opt-in `.aimodel` handling in the `ultralytics_yolo` plugin | 2 (after the `UltralyticsYOLO` 8.9.15 floor bump) |

## Why opt-in and not the default

Measured on an iPhone 17 Pro (iOS 27.0), same YOLO26n weights per task as Core ML INT8 (shipped), Core ML FP16 and Core AI FP16, through the same SDK call. Totals (pre + inference + post, ms) with hardware acceleration; full tables, both heads and the CPU-only path are in [`docs/performance.md`](https://github.com/ultralytics/yolo-ios-app/blob/coreai-default-ios/docs/performance.md#on-device-results).

| Task | Core AI FP16 (official: raw head) | Core ML FP16 | Core ML INT8 (shipped) | Verdict |
| --- | --- | --- | --- | --- |
| Detect | **3.73** (1.15 / 2.05 / 0.54) | 4.05 | 4.59 | Core AI ahead |
| Segment | 5.33 (1.30 / 2.79 / 1.23) | **5.24** | 5.79 | tie |
| Pose | **4.02** (1.40 / 2.31 / 0.31) | 4.06 | 4.09 | tie |
| OBB | **3.65** (1.27 / 1.97 / 0.41) | 3.99 | 3.79 | tie, Core AI slightly ahead |
| Classify | **1.19** (0.68 / 0.51 / 0.01) | 2.34 | 1.98 | Core AI ahead |
| Semantic | 8.55 (1.29 / 3.47 / 3.79) | **4.86** | 4.95 | Core AI behind |
| Depth | 7.87 (1.37 / 5.64 / 0.85) | 6.62 | **6.42** | Core AI behind |

- End to end Core AI is level with Core ML rather than ahead: its model time is about half of Core ML's with the raw head, but about 1.2 ms of CPU preprocessing per frame (ultralytics/yolo-ios-app#321) and the Swift NMS give most of that back. Semantic (ultralytics/yolo-ios-app#322) and depth are slower, and CPU-only inference is 1.5 to 2x slower.
- Core AI assets are FP16 (no INT8 Core AI export yet), about twice the download size of the INT8 Core ML assets.
- The iOS 27.0 runtime has open correctness issues in neighbouring graphs: an FP16 end-to-end pose head (apple/coreai-torch#115) and an in-graph `argmax` (apple/coreai-torch#116) silently return wrong outputs under hardware acceleration, and a stale specialization cache entry fails model loads until it is evicted (handled in ultralytics/yolo-ios-app#319). The official opt-in assets avoid the first two by using the raw head and logits.
- What Core AI does better today: model loads after the first specialization take tens of ms against a 1 to 2 s Core ML compile on every launch, so switching models is visibly faster.

The 35 official opt-in assets (7 tasks x n/s/m/l/x, `format="coreai", quantize=16`, raw head for detect, segment, pose and OBB) are published on the yolo-ios-app [`v8.3.0`](https://github.com/ultralytics/yolo-ios-app/releases/tag/v8.3.0) release as `<model>.aimodel.zip`. All 7 nano tasks, detect s/m/l/x and the x model of every other task were loaded and checked on the device; the other 18 were exported and verified on the Mac only.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary

Added an opt-in Apple Core AI backend for `.aimodel` assets on iOS 27+ devices while keeping Core ML as the default backend.

### 📊 Key Changes

- Added `CoreAIRequest` support with weakly linked, availability-gated Core AI loading, CPU or hardware-accelerated execution, specialization-cache recovery, metadata label loading, and `PredictorError` cases for unavailable or invalid models.
- Unified Core ML and Core AI outputs through `VNRequest` and `featureArrays(_:)`; existing seven task decoders now consume either backend, with Core AI-specific preprocessing, tensor conversion, timing, and classification probability handling.
- Extended model resolution, downloading, extraction, validation, and caching for `.aimodel` files and `.aimodel.zip` URLs, while keeping Core ML and Core AI cache entries separate.
- Added the app’s default-off **Core AI Models (iOS 27+)** Settings toggle, format-specific model lists, foreground reload behavior, and download handling that preserves the selected asset format.
- Updated `export-models.py`, documentation, performance records, tests, and the package/app version to 8.9.15; Core AI exports use FP16 raw-head assets and require explicit `--formats coreai`.

### 🎯 Purpose & Impact

- SDK users can select Core AI by passing an `.aimodel` path or `.aimodel.zip` URL; explicit Core AI models fail with `coreAIUnavailable` on unsupported OS versions, SDKs, or the iOS Simulator.
- App users continue to receive the existing Core ML workflow by default, while iOS 27+ device users can enable Core AI from Settings and switch between independently cached Core ML and Core AI assets.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @glenn-jocher, thank you for submitting a `ultralytics/yolo-ios-app` 🚀 PR! This automated message confirms your contribution was received, and an Ultralytics engineer will assist with the review. To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/yolo-ios-app/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/yolo-ios-app` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

**Comment by UltralyticsAssistant** [comment]
🎉 Huge thanks to @glenn-jocher for the excellent work in [merged PR #319](https://github.com/ultralytics/yolo-ios-app/pull/319)!

> “The only way to do great work is to love what you do.” — Steve Jobs

This update brings opt-in Core AI support for `.aimodel` assets on iOS 27+ while preserving Core ML as the default, with unified outputs, format-aware caching, and robust model handling. Great contribution to the iOS experience!
